### ISSUE TITLE
On a custom time series chart widget, how do I access attributes after setting hasAdditionalLatestDataKeys: true?

### ISSUE THREAD

**Issue by csann** [thingsboard/thingsboard#8741]
**Component**
 * UI

**Description**
Starting with TB version 3.4, it is possible to access attributes in TB supplied standard time series charts. After examining the flot source code, I modified my custom widget source code as follows:

```
self.onDataUpdated = function() {
    console.log(self.ctx.data.length);

    for (var i = 0; i < self.ctx.data.length; i++) {
        var datasourceData = self.ctx.data[i];
        var dataSet = datasourceData.data;
        myChart.data.datasets[i].data.length = 0;
        var data = myChart.data.datasets[i].data;
        console.log(dataSet);
        for (var d = 0; d < dataSet.length; d++) {
            var tsValuePair = dataSet[d];
            var ts = tsValuePair[0];
            var value = tsValuePair[1];
            data.push({
                t: ts,
                y: value
            });
        }
    }

    myChart.options.scales.xAxes[0].ticks.min = self.ctx
        .timeWindow.minTime;
    myChart.options.scales.xAxes[0].ticks.max = self.ctx
        .timeWindow.maxTime;
    myChart.update();
}

self.typeParameters = function() {
    return {
        hasAdditionalLatestDataKeys: true
    };
}
```
Adding "hasAdditionalLatestDataKeys: true" caused the datasource configuration screen to allow me to configure both time series and attribute data sources. 

Unfortunately I cannot see them in the onDataUpdated event, no matter how I configure my data sources. I am concerned this feature is not available for custom time series data charts. Am I correct about this? 

If so, the only other way I know to access attribute data is via attributeService.getEntityAttributes. After a lot of work, I implemented this in my onInit event but it is very difficult to deal with the asynchronous nature of getEntityAttributes. 

Because it takes a second or two for each attribute, I see no way to implement this in the onDataUpdated event and so I do not think this method can be used to respond to real time changes of attributes. 

I have little experience with Thingsboard so I may be wrong about my conclusions.

So, in summary I have the following questions:

1. Is there any way to use hasAdditonalLatestDataKeys in a custom time series chart widget to access the latest attribute values in the onDataUpdatedEvent? 
2. If so, can someone suggest any changes I need to make in my onDataUpdatedEvent code?
3. If not, would Thingsboard consider adding this feature to custom widgets?
4. If not, are they any other ways I can somehow subscribe to attribute changes within my custom widget?

**Environment**

 * OS:  MacOS 13.4
 * ThingsBoard: Cloud (3.5.?)
 * Browser: Safari

**Comment by csann** [comment]
Would someone please help me with this?

**Comment by devdoob** [comment]
Have you added those attirubutes in data source? visually in widget

**Comment by csann** [comment]
Muhammad

Yes I did add the attributes to my custom widget after I put the custom widget in a dashboard. 

The problem I have is that I can’t figure out how to access the attributes in my custom widget code. I need to access them in the onInit and OnDataUpdated functions in order to change certain graphical element of my custom widget. I suspect it may have something to do with the Data Key Settings Schema or the Latest Data Key Settings Schema sections of the custom widget editor. So far I have put nothing in those sections. 

Do you have any idea how to proceed?

Clark

**Comment by devdoob** [comment]
can you please share the screenshots how you added the atributes to the widget?

**Comment by csann** [comment]
Muhammad

Thank you for taking the time to help!

I have added the two attributes I need in two different ways. Both sorta work, but neither allow me to access the data. 

￼

￼


The problem with both methods of adding the attributes, is that both allow me to see the attribute data in the debugger (actually in two places), but in both cases I cannot access the data. Both methods of adding attributes in the dashboard editor product the same data structures in the debugger. H

ere is what the debugger shows:

￼

￼

The data is in two places, one inside the defaultSubscription object, and inside the latestData array. I can see the data in the debugger but I cannot figure out how to access it in my code. 

Here is the code in my onDataUpdated function:

￼

You can see where I tried to access the attributes from the two different places. No matter what I do, the widget editors says they are invalid statements.

You can also see where I added hasAdditionalLatestDataKeys. I understand that is needed to cause time series widgets to also include attributes.

Do you have any recommendations?

Clark



Here is

**Comment by Ultrazombie** [comment]
As the community has not provided assistance on this matter over an extended period, I am closing this ticket due to inactivity.
